gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,139 @@
+#include <dirent.h>
+#include <sys/stat.h>
+
 #include "postgres.h"
 #include "fmgr.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "nodes/execnodes.h"
+#include "cdb/cdbvars.h"
+#include "utils/builtins.h"
+#include "utils/relfilenodemap.h"
+#include "utils/timestamp.h"
 
 PG_MODULE_MAGIC;
+
+#define should_skip(filename)	(pg_strncasecmp(filename, "pg", 2) == 0 \
+								|| pg_strncasecmp(filename, "t_", 2) == 0 \
+								|| pg_strncasecmp(filename, ".", 2) == 0 \
+								|| pg_strncasecmp(filename, "..", 2) == 0 \
+								|| pg_strncasecmp(filename + strlen(filename) - 4, "_fsm", 4) == 0 \
+								|| pg_strncasecmp(filename + strlen(filename) - 3, "_vm", 3) == 0 \
+								|| pg_strncasecmp(filename + strlen(filename) - 5, "_init", 5) == 0)
+
+PG_FUNCTION_INFO_V1(adb_get_relfilenodes);
+Datum adb_get_relfilenodes(PG_FUNCTION_ARGS)
+{
+	Oid datoid = MyDatabaseId;
+	Oid tablespace_oid = PG_GETARG_OID(0);
+
+	struct dirent *direntry;
+	DIR *dirdesc;
+	char filename[MAXPGPATH * 2];
+	char *datpath;
+
+	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc tupdesc;
+	Tuplestorestate *tupstore;
```

Comment:
Why did you choose the materialization of the results (tuplestore), and not the line-by-line output (ValuePerCall, SRF_IS_FIRSTCALL, SRF_RETURN_NEXT, ...)?